picard/file.py

```diff
@@ -167,7 +167,15 @@ def _load_check(self, filename):
         if self.tagger.stopping:
             log.debug("File not loaded because %s is stopping: %r", PICARD_APP_NAME, self.filename)
             return None
-        return self._load(filename)
+        try:
+            # Try loading based on extension first
+            ret = self._load(filename)
+        except Exception:
+            from picard.formats import guess_format
+            # If it fails, force guess and try loading again
+            self = guess_format(filename)
```

Comment:
A small improvement could be that it just raises if the newly detected format is the same as before. Also guess_format might return `None`. So maybe something like this:

```
format = guess_format(filename)
if not format or format == self:
    raise
ret = format._load(filename)
```

Reply:
Do we need more unit tests for the guessing functionality?

Reply:
Hmm, the score function checks for the extension. I'm removing that.

Reply:
That's likely because mutagen itself also considers the extension in its scoring logic, so it is not a pure content based detection. Our own extension based check still is a performance shortcut :)

Reply:
I don't think you need a workaround. Let's just leave it as is for now. I don't think it does much harm that mutagen additionally also considers the extension. If this is something to change it should be done on mutagen.

Also it's mostly unrelated to this change, if we want to update the scoring logic from the Picard side we should do this in a separate PR.

Reply:
Now it's calling guess_format twice ;) You should store it in a variable. Also you can do the comparison with self then. No need to try loading with the same format twice (e.g. an broken MP3 file that gets correctly detected by extension and content but which will not load due to a different error, no need to open it twice)

Reply:
Saving it to a variable and then assigning throws an exception, but doing it directly to self doesn't. I don't think it's a huge problem for three reasons: you don't really expect files with wrong extensions that often (e.g. VLC completely freaks out), files without extensions were already guessed and the files are already cached in memory and will have their metadata loaded in the near future. It is quite different from guessing every single file format, and then trying to load each of them, having the file cache purged in the meantime.

Reply:
Measured the loading times: fast path ~1ms per file, slow path ~2ms per file. The difference between using one or two guess_format is statistically irrelevant.

Reply:
> Saving it to a variable and then assigning throws an exception, but doing it directly to self doesn't.

> The difference between using one or two guess_format is statistically irrelevant, as expected.

I don't get why calling `guess_format()` twice is needed. Even if performance isn't much impacted, what prevent us to get rid of a useless call?

Reply:
In this case, my mistake. It works fine if the guessed file is correct and fails when it doesn't (e.g. rename an .aac as .mp3, due to how mutagen scores are calculated it will be recognized as an MP3 file, and will fail to load. Removing the duplicated call.